content-scripts/cs.js

```diff
@@ -89,43 +89,48 @@ chrome.runtime.onMessage.addListener((request, sender, sendResponse) => {
   console.log("[Message from background]", request);
   if (request === "getInitialUrl") {
     sendResponse(initialUrl);
-  } else if (request.themesUpdated) {
-    injectUserstylesAndThemes({ themes: request.themesUpdated, isUpdate: true });
   }
 });
 
 // Store all themes that were enabled this session
 const sessionEnabledThemes = new Set();
 
-function injectUserstylesAndThemes({ addonsWithUserstyles = [], themes, isUpdate }) {
-  document.querySelectorAll(".scratch-addons-theme").forEach((style) => {
-    if (!style.textContent.startsWith("/* sa-autoupdate-theme-ignore */")) style.remove();
-  });
-  const userstyles = addonsWithUserstyles.map((addon) => addon.styles);
-  for (const addon of userstyles || []) {
-    for (const userstyle of addon) {
+function addStyle(addon) {
+  for (let userstyle of addon.styles) {
+    if (addon.injectAsStyleElt) {
+      sessionEnabledThemes.add(addon.addonId);
+
+      // Replace %addon-self-dir% for relative URLs
+      userstyle = userstyle.replace(/\%addon-self-dir\%/g, chrome.runtime.getURL(`addons/${addon.addonId}`));
+      //userstyle += `\n/*# sourceURL=${styleUrl} */`;
+      const style = document.createElement("style");
+      style.classList.add("scratch-addons-style");
+      style.setAttribute("data-addon-id", addon.addonId);
+      style.setAttribute("data-addon-index", addon.index);
+      style.textContent = userstyle;
+
+      if (document.body) document.documentElement.insertBefore(style, document.body);
+      else document.documentElement.appendChild(style);
+    } else {
       const link = document.createElement("link");
       link.rel = "stylesheet";
-      link.href = userstyle.url;
+      link.setAttribute("data-addon-id", addon.addonId);
+      link.classList.add("scratch-addons-style");
+
+      link.href = userstyle;
       if (document.body) document.documentElement.insertBefore(link, document.body);
       else document.documentElement.appendChild(link);
     }
   }
-  for (const theme of themes) {
-    sessionEnabledThemes.add(theme.addonId);
-    for (const styleUrl of theme.styleUrls) {
-      let css = theme.styles[styleUrl];
-      // Replace %addon-self-dir% for relative URLs
-      css = css.replace(/\%addon-self-dir\%/g, chrome.runtime.getURL(`addons/${theme.addonId}`));
-      if (isUpdate && css.startsWith("/* sa-autoupdate-theme-ignore */")) continue;
-      css += `\n/*# sourceURL=${styleUrl} */`;
-      const style = document.createElement("style");
-      style.classList.add("scratch-addons-theme");
-      style.setAttribute("data-addon-id", theme.addonId);
-      style.textContent = css;
-      if (document.body) document.documentElement.insertBefore(style, document.body);
-      else document.documentElement.appendChild(style);
-    }
+}
+function removeAddonStyles(addonId) {
+  document.querySelectorAll(`[data-addon-id='${addonId}']`).forEach((style) => style.remove());
```

Comment:
Alternatively, we could set `.disabled = true`, works both for style and link elements. Not sure if it's worth doing, tho.